File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/storage/StorageBlock.java
package com.github.worldoneo.ciphelet.storage;

import android.util.Base64;

import com.github.worldoneo.ciphelet.connector.Connector;

import javax.crypto.spec.IvParameterSpec;

public class StorageBlock {
    public final String json;
    public final IvParameterSpec iv;

    public StorageBlock(String json, IvParameterSpec iv) {
        this.json = json;
        this.iv = iv;
    }

    private final class SerializeableBlock {
        public final String iv;
        public final String json;

        private SerializeableBlock(String iv, String json) {
            this.iv = iv;
            this.json = json;
        }

        public String serialize() {
            return Connector.gson.toJson(this);
        }
    }

    public String serialize() {
        String iv = Base64.encodeToString(this.iv.getIV(), Base64.URL_SAFE);
        return Connector.gson.toJson(new SerializeableBlock(iv, this.json));
    }

    public static StorageBlock deserialize(String json)  {
        SerializeableBlock serializeableBlock = Connector.gson.fromJson(json, SerializeableBlock.class);
        byte[] iv = Base64.decode(serializeableBlock.iv, Base64.URL_SAFE);
        return new StorageBlock(serializeableBlock.json, new IvParameterSpec(iv));
    }
}

File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/storage/StorageBlock.kt
package com.github.worldoneo.ciphelet.storage

import android.util.Base64
import com.github.worldoneo.ciphelet.connector.Connector.Companion.gson
import javax.crypto.spec.IvParameterSpec

class StorageBlock(val json: String, val iv: IvParameterSpec) {

    private inner class SerializeableBlock(val iv: String, val json: String) {
        fun serialize(): String {
            return gson.toJson(this)
        }

    }

    fun serialize(): String {
        val iv = Base64.encodeToString(iv.iv, Base64.URL_SAFE)
        return gson.toJson(SerializeableBlock(iv, json))
    }

    companion object {
        fun deserialize(json: String?): StorageBlock {
            val serializeableBlock = gson.fromJson(json, SerializeableBlock::class.java)
            val iv = Base64.decode(serializeableBlock.iv, Base64.URL_SAFE)
            return StorageBlock(serializeableBlock.json, IvParameterSpec(iv))
        }
    }

}